### PR TITLE
Batch sn payments

### PR DESCRIPTION
**Pull request by darcys22**

### CODE REVIEW

**Comment by majestrate** [comment]
this will be a chonky PR when it's ready

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 1]
```diff
@@ -105,7 +105,7 @@ bool BlockchainSQLite::subtract_sn_payments(cryptonote::network_type nettype, st
     if (cryptonote::is_valid_address(std::get<0>(payment), nettype))
     {
       auto prev_entry = m_storage->get<cryptonote::batch_sn_payments>(std::get<0>(payment));
-      if ((prev_entry.amount - std::get<1>(payment)) < 0)
+      if ((int(prev_entry.amount) - std::get<1>(payment)) < 0)
```
Comment (majestrate) [inline src/blockchain_db/sqlite/db_sqlite.cpp:108]: pedantic style gripe:

`int{prev_entry.amount}`

Comment (darcys22) [reply src/blockchain_db/sqlite/db_sqlite.cpp:108]: Cheers will do

File: src/cryptonote_core/blockchain.cpp [hunk 2]
```diff
@@ -5357,24 +5367,31 @@ bool Blockchain::prepare_handle_incoming_blocks(const std::vector<block_complete
...
       // get all amounts from tx.vin(s)
       for (const auto &txin : tx.vin)
       {
-        const auto& in_to_key = var::get<txin_to_key>(txin);
 
-        // check for duplicate
-        auto it = its->second.find(in_to_key.k_image);
-        if (it != its->second.end())
-          SCAN_TABLE_QUIT("Duplicate key_image found from incoming blocks.");
+        if (!std::holds_alternative<txin_gen>(txin))
```
Comment (majestrate) [inline src/cryptonote_core/blockchain.cpp:5375]: shouldn't this be `if(std::holds_alternative<tx_to_key>(txin))` ?

Comment (darcys22) [reply src/cryptonote_core/blockchain.cpp:5375]: Yeah it should, and there are a few times I have done this. My reasoning is that some of these code blocks (maybe not this one) have `assert input = <txin_to_key>` and wanted those asserts to still trigger if a `txin_to_script` or something weird came through. I don't think `txin_to_scripts` are used here but was unsure, so temporarily just allowing `txin_gen` to exist here with a mental note to lock down what types of inputs can go into a transaction.

**Comment by darcys22** [comment]
Todo:

- [ ] Tests
- [ ] switch sqlite_orm to another lib
- [ ] use trigger to clear out zero amount balances
- [ ] constrain sqlite table to only allow positive balances
- [ ] check functionality of subaddresses

**Trigger**
```
CREATE TRIGGER batch_payments_delete_empty
AFTER UPDATE ON batch_payments FOR EACH ROW WHEN NEW.amount = 0 THEN DELETE FROM batch_payments WHERE address = NEW.address;
```
**Constraint**
```
CHECK(amount >= 0)
```

**Comment by jagerman** [comment]
> switch sqlite_orm to another lib

-> https://github.com/SRombauts/SQLiteCpp now used in storage server, lets you write your own queries but keeps a C++ interface that's much nicer to work with than the straight C interface.

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 3]
```diff
@@ -147,12 +156,57 @@ void BlockchainSQLite::load_database(std::optional<fs::path> file)
...
+  }
+}
+
+bool BlockchainSQLite::increment_height() {
+  try {
+    SQLite::Transaction transaction{*db};
+    LOG_PRINT_L3("BlockchainDB_SQLITE::" << __func__);
+    this->height++;
+    SQLite::Statement update_height{*db,
+      "UPDATE batch_sn_info SET height = ?"};
```
Comment (majestrate) [inline src/blockchain_db/sqlite/db_sqlite.cpp:175]: I think that `UPDATE batch_sn_info SET height = height + 1`  would work here if you are always just incrementing the value by 1.

Comment (darcys22) [reply src/blockchain_db/sqlite/db_sqlite.cpp:175]: Yeah 100%, using the parameter is redundant in hindsight. 

What are your thoughts on calling the update here? This will get called after every block is added and im concerned it will be too slow, maybe updating the database height record in the destructor of my SQLite struct? Although that possibly will cause sync issues on an unexpected crash

File: cmake/CMakeDoxyfile.in [hunk 4]
```diff
@@ -0,0 +1,280 @@
+#
+# DO NOT EDIT! THIS FILE WAS GENERATED BY CMAKE!
+#
```
Comment (darcys22) [inline cmake/CMakeDoxyfile.in:3]: @majestrate Are these CMake files necessary? I'm unsure how i've generated so many and whether they should even be committed to this PR

Comment (majestrate) [reply cmake/CMakeDoxyfile.in:3]: probably best to exclude them

File: src/blockchain_db/sqlite/db_sqlite.cpp (hunk 3 above)
Comment (majestrate) [reply src/blockchain_db/sqlite/db_sqlite.cpp:175]: you could probably recompute expected height on startup to mitigate inconsistent state from crashes.

**Comment by darcys22** [comment]
Misc other issues:

- [ ] Clean up CMake folder
- [ ] Rebase
- [ ] Simulate batching and decide on final length of batching period

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 5]
```diff
@@ -97,18 +97,20 @@ int exec_query(SQLite::Database& db, const char* query, const T&... bind) {
...
+  else
+    filename = "file:memdb" + (int{filename.at(10)} + 1) + std::string("?mode=memory");
 
   this->load_database(filename);
-  this->create_schema();
 
   std::vector<std::tuple<std::string, int64_t, int64_t>> all_payments;
 
-  SQLite::Statement st{*db, "SELECT * FROM batch_sn_payments"};
+  SQLite::Statement st{*other.db, "SELECT * FROM batch_sn_payments"};
```
Comment (majestrate) [inline src/blockchain_db/sqlite/db_sqlite.cpp:113]: `db->other` ?

File: CMakeLists.txt [hunk 6]
```diff
@@ -86,6 +86,9 @@ else()
   message(WARNING "PIE disabled: cmake 3.14+ is required for proper PIE linking support")
 endif()
 
+# TODO sean remove this
+set(CMAKE_CXX_LINK_EXECUTABLE "${CMAKE_CXX_LINK_EXECUTABLE} -ldl")
```
Comment (darcys22) [inline CMakeLists.txt:90]: This was something that came up when trying to run cmake with `-DCMAKE_BUILD_TYPE=Debug` this made it work but unsure exactly what is going on behind the scenes and whether it should stay here or not

File: external/CMakeLists.txt [hunk 7]
```diff
@@ -143,3 +145,47 @@ target_link_libraries(cpr PUBLIC CURL::libcurl)
...
+#set(SQLITECPP_RUN_CPPCHECK OFF CACHE BOOL "" FORCE)
+#set(SQLITECPP_RUN_DOXYGEN OFF CACHE BOOL "" FORCE)
+#set(SQLITECPP_BUILD_EXAMPLES OFF CACHE BOOL "" FORCE)
+#set(SQLITECPP_BUILD_TESTS OFF CACHE BOOL "" FORCE)
+
+#add_subdirectory(SQLiteCpp EXCLUDE_FROM_ALL)
+
+
+#pkg_check_modules(SQLite3 sqlite3>=3.35.5 IMPORTED_TARGET)
+
```
Comment (darcys22) [inline external/CMakeLists.txt:179]: Still having issues compiling our static builds. The uncommented lines below work for debian and locally building static.

File: src/cryptonote_config.h [hunk 8]
```diff
@@ -257,6 +257,11 @@ namespace config
   inline constexpr std::string_view HASH_KEY_CLSAG_AGG_0 = "CLSAG_agg_0"sv;
   inline constexpr std::string_view HASH_KEY_CLSAG_AGG_1 = "CLSAG_agg_1"sv;
 
+  //Batching SN Rewards
+  inline constexpr uint64_t BATCHING_INTERVAL = 6;
```
Comment (darcys22) [inline src/cryptonote_config.h:261]: TODO more testing to decide what the optimal length for this batching interval is. Preliminary python scripts were indicating that 200 possibly is a good length for a 40% reduction in block outputs. 2000 will get us an 80% reduction however I have some concerns about coinbase payments having too many outputs in a single block if we do that high.

File: tests/core_tests/oxen_tests.cpp [hunk 9]
```diff
@@ -3325,3 +3354,276 @@ bool oxen_pulse_chain_split_with_no_checkpoints::generate(std::vector<test_event
...
+    std::vector<cryptonote::block> blocks;
+    bool r = c.get_blocks((uint64_t)0, (uint64_t)-1, blocks);
+    CHECK_TEST_CONDITION(r);
+    std::vector<cryptonote::block> chain;
+    map_hash2tx_t mtx;
+    r = find_block_chain(events, chain, mtx, cryptonote::get_block_hash(blocks.back()));
+    CHECK_TEST_CONDITION(r);
+
+    // Expect increase in balance after blocks < batched constant
+    // 201 (balance) - 100 (stake) - 0.2 (test fee) + 16.5*Batching_Interval (Batched reward)
```
Comment (darcys22) [inline tests/core_tests/oxen_tests.cpp:3452]: Unsure why this is 201

**Comment by darcys22** [comment]
This is now ready for review. 

- The macos tests are failing and are awaiting an upstream change which should resolve the issue. https://github.com/SRombauts/SQLiteCpp/pull/335
- There `oxen_service_nodes_test_rollback` test fails but is inconsistent, went over the codebase with Valgrind to see if I could find a memory leak/uninitialised variable which resulted in #1482 however I'm not sure if this was the root cause. Will have another go searching for this 
- Static builds are failing but I've exhausted my knowledge trying to get them to pass and need some help. CMake is hard
- Discussion on optimal batching interval size to be scheduled

File: src/blockchain_db/sqlite/db_sqlite.h [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+  std::vector<cryptonote::batch_sn_payment> get_block_payments(cryptonote::network_type nettype, uint64_t block_height);
+  bool delete_block_payments(uint64_t block_height);
+
+  uint64_t height;
+
+  std::unique_ptr<SQLite::Database> db;
+  std::string filename;
+
+
+private:
```
[the remark this answers is not in the document]
Comment (jagerman) [reply src/blockchain_db/sqlite/db_sqlite.h:94]: If test code needs to get at these, one minimally intrusive approach is to use something like this in the test suite that exposes the internals:
```C++
class TestingBlockchainSQLite : public BlockchainSQLite {
public:
    using BlockchainSQLite::BlockchainSQLite; // inherit constructor
    
    uint64_t testing_get_height() { return height; }
    uint64_t batching_count() { ... } // I think this one currently is only for the test suite
}
```
and then add only a `friend class TestingBlockchainSQLite;` in the main class, keeping the actual exposing code limited to the testing subclass.

(The copy constructor code that duplicates the DB would be another thing to confine to this subclass, as a named function, assuming that is also meant for test suite code).

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 11]
```diff
@@ -0,0 +1,662 @@
...
+  st.bind(1, address);
+  std::optional<uint64_t> amount = std::nullopt;
+  while (st.executeStep()) {
+    assert(!amount);
+    amount.emplace(st.getColumn(0).getInt64());
+  }
+  return amount;
+}
+
+bool BlockchainSQLite::add_sn_payments(cryptonote::network_type nettype, std::vector<cryptonote::batch_sn_payment>& payments, uint64_t block_height)
```
Comment (jagerman) [inline src/blockchain_db/sqlite/db_sqlite.cpp:290]: Could we maybe specify this network_type in the constructor (and store it as a member) so that we don't have to pass it in these various functions?  You can't ever change it, so this is always passing in the same value.

**Review by jagerman — COMMENTED** [review]
First pass, left many comments/suggestions, some tiny, some bigger.

More to come.

**Comment by darcys22** [comment]
So redid the simulation for the batching and got the following results:

batched_savings

Source code can be seen here:
https://github.com/darcys22/batched-sn-rewards-simulation

Results indicate that we get a logarithmic curve with steps over an interval of what appears to be the total service node count (Data is from a slightly outdated service node list). Likely a 70% saving in outputs for a batching interval 5% less than total service node list and 80% saving for 5% above.

If the batching interval was equal to 1800 (approx total service nodes at time of writing) we would have a payout every 2.5 days is probably still reasonable for service node operators.

Also tested the outputs per block and looks like its a constant equal to the max number of contributors to a service node.

Looking at this data I'd probably recommend setting the batching interval to an easy to remember 2000 blocks.

**Comment by darcys22** [comment]
```
curl -X POST http://public.loki.foundation:38157/json_rpc -d '{"jsonrpc":"2.0","id":"0","method":"get_block_header_by_height","params":{"height":674299}}' | jq .
```
Should respond with sn winner and reward in block header

**Comment by jagerman** [comment]
> Should respond with sn winner and reward in block header

Also: `get_block_header_by_hash` likely has the same issue (and I'd hope the same root cause).

**Review by Lucifer1903 — CHANGES_REQUESTED** [review]
I would like to suggest setting the batching interval to 2520 (3.5 days). This would make payments happen twice a week at the same day and time every week.

In my opinion that would be way easier for people to keep track of rather than the payment day changing week by week.

**Comment by jagerman** [comment]
> I would like to suggest setting the batching interval to 2520 (3.5 days). This would make payments happen twice a week at the same day and time every week.
> 
> In my opinion that would be way easier for people to keep track of rather than the payment day changing week by week.

Sean and I had a talk about this: one obstacle is that the current implementation essentially starts your N-block window when you earn a first reward when you previously had no balance.  So if you got paid out at block 5000, then earned your next reward at 5555, your next payout would come at 7555, not 7000 (assuming 2000 block window).  That introduces some variation into the reward schedule, as the 2000 essentially becomes a lower-bound on the payment time (the upper bound would be 2000 + #-of-SNs).

I think, however, we might be able to address that--perhaps by keeping the row in place with a balance of 0 and only pruning them from the database once they get noticeably more than N blocks old (e.g. more than 2N).

**Comment by darcys22** [comment]
Action items for this batching. Changing the accrual of fees to drip feed rewards to all SNs per block. So instead of earning a full reward in a ring. Every block will reward all nodes a small amount `full_reward/n_service_nodes`.

One change it will need is a rule to the effect of:
- if your service node has not been an active service node for 1 day (720 blocks) then you don't get block rewards (but you would still earn TX fees if you are the block proposer).
So new registrations + decomm/recomm + ip penalties would all pause you from rewards for 1 day.

The reward calculation would become the small constant amount added to all sns + block proposer gets transaction fees. And the block proposer is the SN recorded in the block so that nodes can validate the fee by referring to their SN list and the block

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 12]
```diff
@@ -0,0 +1,647 @@
...
+#include "cryptonote_basic/hardfork.h"
+
+#undef OXEN_DEFAULT_LOG_CATEGORY
+#define OXEN_DEFAULT_LOG_CATEGORY "blockchain.db.sqlite"
+
+namespace cryptonote
+{
+
+BlockchainSQLite::BlockchainSQLite(cryptonote::network_type nettype, fs::path db_path)
+  : db::Database(db_path, ""), m_nettype(nettype), filename(db_path.u8string())
```
Comment (majestrate) [inline src/blockchain_db/sqlite/db_sqlite.cpp:50]: use curly braces for initialization of members.

File: src/blockchain_db/sqlite/db_sqlite.h [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+
+namespace cryptonote
+{
+
+fs::path check_if_copy_filename(std::string_view db_path);
+
+class BlockchainSQLite : public db::Database
+{
+public:
+  BlockchainSQLite(cryptonote::network_type nettype, fs::path db_path);
```
Comment (majestrate) [inline src/blockchain_db/sqlite/db_sqlite.h:49]: explicit?

File: src/blockchain_db/sqlite/db_sqlite.h [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+  cryptonote::network_type m_nettype;
+  std::string filename;
+
+};
+
+class BlockchainSQLiteTest : public BlockchainSQLite
+{
+public:
+  BlockchainSQLiteTest(cryptonote::network_type nettype, fs::path db_path)
+    : BlockchainSQLite(nettype, db_path) {};
```
Comment (majestrate) [inline src/blockchain_db/sqlite/db_sqlite.h:103]: can we have this constructor declared only in the header and then implemented in a source file?

File: src/cryptonote_basic/cryptonote_format_utils.cpp [hunk 15]
```diff
@@ -55,11 +55,6 @@
 
 using namespace crypto;
 
-static std::atomic<uint64_t> tx_hashes_calculated_count(0);
```
Comment (majestrate) [inline src/cryptonote_basic/cryptonote_format_utils.cpp:58]: @tewinget

File: src/blockchain_db/CMakeLists.txt [hunk 16]
```diff
@@ -30,9 +30,12 @@
 add_library(blockchain_db
   blockchain_db.cpp
   lmdb/db_lmdb.cpp
+  sqlite/db_sqlite.cpp
   )
 
 target_link_libraries(blockchain_db
+  PUBLIC
+    SQLiteCpp
```
[the remark this answers is not in the document]
Comment (tewinget) [reply src/blockchain_db/CMakeLists.txt:38]: note: revisit

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 17]
```diff
@@ -0,0 +1,666 @@
...
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "db_sqlite.h"
+
+#include <sodium.h>
+#include <SQLiteCpp/SQLiteCpp.h>
```
Comment (tewinget) [inline src/blockchain_db/sqlite/db_sqlite.cpp:31]: Ideally this dep is supplanted by `src/sqlitedb/...`

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 18]
```diff
@@ -0,0 +1,666 @@
...
+    create_schema();
+  }
+
+  SQLite::Statement st{db, "SELECT height FROM batch_db_info"};
+  while (st.executeStep()) {
+    this->height = st.getColumn(0).getInt64();
+  }
+}
+
+void BlockchainSQLite::create_schema() {
```
Comment (tewinget) [inline src/blockchain_db/sqlite/db_sqlite.cpp:65]: TODO: metadata table with version info etc.

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 19]
```diff
@@ -0,0 +1,666 @@
...
+namespace cryptonote
+{
+
+BlockchainSQLite::BlockchainSQLite(cryptonote::network_type nettype, fs::path db_path)
+  : db::Database(db_path, ""), m_nettype(nettype), filename(db_path.u8string())
+{
+  LOG_PRINT_L3("BlockchainDB_SQLITE::" << __func__);
+  height = 0;
+
+  if (!db.tableExists("batched_payments_accrued") || !db.tableExists("batched_payments_raw") || !db.tableExists("batch_db_info")) {
```
Comment (tewinget) [inline src/blockchain_db/sqlite/db_sqlite.cpp:55]: add metadata table, then just check for it?

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 20]
```diff
@@ -0,0 +1,666 @@
...
+    address VARCHAR NOT NULL,
+    amount BIGINT NOT NULL,
+    height_paid BIGINT,
+    PRIMARY KEY(address, height_paid),
+    CHECK(amount >= 0)
+);
+
+CREATE INDEX batched_payments_raw_height_idx ON batched_payments_raw (height_paid);
+
+CREATE TABLE batch_db_info (
```
Comment (tewinget) [inline src/blockchain_db/sqlite/db_sqlite.cpp:93]: should be part of a "metadata" table with one row

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 21]
```diff
@@ -0,0 +1,666 @@
...
+    height BIGINT NOT NULL
+);
+
+INSERT INTO batch_db_info (height) VALUES (0);
+
+CREATE TRIGGER batch_payments_prune
+AFTER UPDATE ON batch_db_info
+FOR EACH ROW
+BEGIN
+    DELETE FROM batched_payments_raw WHERE height_paid < (NEW.height - 10000);
```
Comment (tewinget) [inline src/blockchain_db/sqlite/db_sqlite.cpp:103]: hard-coded const of 10000, should be parametrized.

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 22]
```diff
@@ -0,0 +1,666 @@
...
+AFTER UPDATE ON batched_payments_accrued
+FOR EACH ROW WHEN NEW.amount = 0
+BEGIN
+    DELETE FROM batched_payments_accrued WHERE address = NEW.address;
+END;
+
+CREATE TABLE batched_payments_raw (
+    address VARCHAR NOT NULL,
+    amount BIGINT NOT NULL,
+    height_paid BIGINT,
```
Comment (tewinget) [inline src/blockchain_db/sqlite/db_sqlite.cpp:86]: should be NOT NULL

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 23]
```diff
@@ -0,0 +1,666 @@
...
+    MTRACE("Adding record for SN reward contributor " << address_str << "to database with amount " << static_cast<int64_t>(payment.amount));
+
+    db::exec_query(insert_payment, address_str, static_cast<int64_t>(payment.amount * 1000));
+    insert_payment.reset();
+  };
+
+  return true;
+}
+
+bool BlockchainSQLite::subtract_sn_payments(std::vector<cryptonote::batch_sn_payment>& payments)
```
Comment (tewinget) [inline src/blockchain_db/sqlite/db_sqlite.cpp:188]: this function's indentation is a bit nuts...`clang-format` it perhaps?

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 24]
```diff
@@ -0,0 +1,666 @@
...
+  return payments;
+}
+
+std::vector<cryptonote::batch_sn_payment> BlockchainSQLite::calculate_rewards(uint8_t hf_version, uint64_t distribution_amount, service_nodes::service_node_info sn_info)
+{
+  LOG_PRINT_L3("BlockchainDB_SQLITE::" << __func__);
+
+  // Find out how much is due for the operator
+  uint64_t operator_fee = 0;
+  {
```
Comment (tewinget) [inline src/blockchain_db/sqlite/db_sqlite.cpp:254]: should be its own function; either in anonymous namespace here or elsewhere if this logic is used elsewhere.

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 25]
```diff
@@ -0,0 +1,666 @@
...
+    if (resulthi > 0)
+      throw std::logic_error("overflow from calculating sn contributor reward");
+    if (resultlo > 0)
+      payments.emplace_back(contributor.address, resultlo, m_nettype);
+  }
+
+  return payments;
+}
+
+bool BlockchainSQLite::add_block(const cryptonote::block& block, const service_nodes::service_node_list::state_t& service_nodes_state)
```
Comment (tewinget) [inline src/blockchain_db/sqlite/db_sqlite.cpp:286]: should return void and `throw` on exception

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 26]
```diff
@@ -0,0 +1,666 @@
...
+  auto fork_height = cryptonote::get_hard_fork_heights(m_nettype, cryptonote::network_version_19);
+  MDEBUG(__FILE__ << ":" << __LINE__ << " (" << __func__ << ") TODO sean remove this - fork height " << (int64_t)fork_height.first.value_or(0) << " - debug");
+  if (block_height == fork_height.first.value_or(0))
+    {
+      MDEBUG(__FILE__ << ":" << __LINE__ << " (" << __func__ << ") TODO sean remove this - " << "Clearing database because block height is on hardfork" << " - debug");
+      clear_database();
+    }
+
+
+  if (block_height != height + 1)
```
Comment (tewinget) [inline src/blockchain_db/sqlite/db_sqlite.cpp:320]: this check should be earlier, and `throw`

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 27]
```diff
@@ -0,0 +1,666 @@
...
+    finalised_payments.emplace_back(calculated_payments_from_batching_db[vout_index].address, std::get<1>(vout), m_nettype);
+    vout_index++;
+  }
+  if (total_oxen_payout_in_vouts != total_oxen_payout_in_our_db)
+  {
+    MERROR("Total service node reward amount incorrect. Should be " << cryptonote::print_money(total_oxen_payout_in_our_db) << ", is: " << cryptonote::print_money(total_oxen_payout_in_vouts));
+    return false;
+  }
+
+  return save_payments(block_height, finalised_payments);
```
Comment (tewinget) [inline src/blockchain_db/sqlite/db_sqlite.cpp:538]: should return `true` and wherever you call this should call `save_payments` itself

Comment (tewinget) [reply src/blockchain_db/sqlite/db_sqlite.cpp:538]: Ideally only the `save` part exists in this db layer, and the validation and calculation logic is handled separately.

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 28]
```diff
@@ -0,0 +1,666 @@
...
+  try
+  {
+    SQLite::Transaction transaction{db, SQLite::TransactionBehavior::IMMEDIATE};
+
+
+    // Step 1: Remove the block producers txn fees
+    uint64_t service_node_reward = cryptonote::service_node_reward_formula(0, block.major_version);
+    if (block.reward - service_node_reward > 0)
+    {
+      if (block.service_node_winner_key && crypto_core_ed25519_is_valid_point(reinterpret_cast<const unsigned char *>(block.service_node_winner_key.data)))
```
Comment (tewinget) [inline src/blockchain_db/sqlite/db_sqlite.cpp:436]: the `is_valid_point` check here is super redundant; if we're *popping* a block where that would be false, how was it added to begin with?

Comment (tewinget) [inline src/blockchain_db/sqlite/db_sqlite.cpp:436]: note: double check `crypto::public_key` for `operator bool` to make sure this `if` does what is expected.

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 29]
```diff
@@ -0,0 +1,666 @@
...
+bool BlockchainSQLite::pop_block(const cryptonote::block& block, const service_nodes::service_node_list::state_t& service_nodes_state)
+{
+  auto block_height = get_block_height(block);
+
+  LOG_PRINT_L3("BlockchainDB_SQLITE::" << __func__ << " called on height: " << block_height);
+  if (height < block_height) {
+    MDEBUG("Block above batching DB height skipping pop");
+    return true;
+  }
+  if (cryptonote::get_block_height(block) != height)
```
Comment (tewinget) [inline src/blockchain_db/sqlite/db_sqlite.cpp:414]: you already have `block_height`

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 30]
```diff
@@ -0,0 +1,666 @@
...
+    INSERT INTO batched_payments_accrued
+        (address, amount) VALUES (OLD.address, OLD.amount)
+        ON CONFLICT (address) DO UPDATE SET amount = (amount + excluded.amount);
+END;
+	)");
+
+	MDEBUG("Database setup complete");
+}
+
+void BlockchainSQLite::clear_database()
```
Comment (jagerman) [inline src/blockchain_db/sqlite/db_sqlite.cpp:129]: I think this should be renamed to `reset_database()` because it's doing a bit more than just clearing it.

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 31]
```diff
@@ -0,0 +1,666 @@
...
+    if (height > block_height)
+    {
+      MDEBUG(__FILE__ << ":" << __LINE__ << " (" << __func__ << ") TODO sean remove this - " << "Clearing database because db height is above block height and pre hardfork" << " - debug");
+      clear_database();
+    }
+    return update_height(block_height);
+  }
+  auto fork_height = cryptonote::get_hard_fork_heights(m_nettype, cryptonote::network_version_19);
+  MDEBUG(__FILE__ << ":" << __LINE__ << " (" << __func__ << ") TODO sean remove this - fork height " << (int64_t)fork_height.first.value_or(0) << " - debug");
+  if (block_height == fork_height.first.value_or(0))
```
Comment (tewinget) [inline src/blockchain_db/sqlite/db_sqlite.cpp:313]: afaict this is the only one of these height checks that is necessary, and the only time that `clear_database` should be necessary as well.  Will revisit.

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 32]
```diff
@@ -0,0 +1,666 @@
...
+  auto hf_version = block.major_version;
+  if (hf_version < cryptonote::network_version_19)
+  {
+    if (height > block_height)
+    {
+      MDEBUG(__FILE__ << ":" << __LINE__ << " (" << __func__ << ") TODO sean remove this - " << "Clearing database because db height is above block height and pre hardfork" << " - debug");
+      clear_database();
+    }
+    return update_height(block_height);
+  }
```
Comment (jagerman) [inline src/blockchain_db/sqlite/db_sqlite.cpp:310]: None of this chunk of code seems necessary.  I think these initial checks can be simplified down to:
- if block height < hard fork height, do nothing (i.e. just return).
- if block height == hard fork height, clear/reset database.

I think everything else will be hf-version-guarded so that it doesn't even look at the database if the block height isn't up to the hard fork height.  So then, up to the hard fork height, we'll have the database in some indeterminate state: it might exist with bogus data, or it might not exist at all, but who cares: because as soon as we reach the hard fork height we're going to nuke it (5 lines after this comment) anyway.

File: src/blockchain_db/sqlite/db_sqlite.cpp (hunk 26 above)
Comment (tewinget) [reply src/blockchain_db/sqlite/db_sqlite.cpp:320]: actually, the sqlite schema could handle this conditional.

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 33]
```diff
@@ -0,0 +1,666 @@
...
+}
+
+bool BlockchainSQLite::update_height(uint64_t new_height)
+{
+  LOG_PRINT_L3("BlockchainDB_SQLITE::" << __func__ << " Called with new height: " << new_height);
+  this->height = new_height;
+  SQLite::Statement update_height{db,
+    "UPDATE batch_db_info SET height = ?"};
+  db::exec_query(update_height, static_cast<int64_t>(height));
+  return true;
```
Comment (jagerman) [inline src/blockchain_db/sqlite/db_sqlite.cpp:155]: No need to return a value here if the value is always the same.

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 34]
```diff
@@ -0,0 +1,666 @@
...
+  auto calculated_rewards = get_sn_payments(block_height);
+
+  // We iterate through the block's coinbase payments and build a copy of our own list of the payments
+  // miner_tx_vouts this will be compared against calculated_rewards and if they match we know the block is
+  // paying the correct people only.
+  std::vector<std::tuple<crypto::public_key, uint64_t>> miner_tx_vouts;
+  for(auto & vout : block.miner_tx.vout)
+    miner_tx_vouts.emplace_back(var::get<txout_to_key>(vout.target).key, vout.amount);
+
+  bool success = false;
```
Comment (tewinget) [inline src/blockchain_db/sqlite/db_sqlite.cpp:337]: this variable *looks* like it's used, but it actually isn't.

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 35]
```diff
@@ -0,0 +1,666 @@
...
+    return false;
+  }
+
+  const auto& conf = get_config(m_nettype);
+  auto hf_version = block.major_version;
+  if (hf_version < cryptonote::network_version_19)
+    return decrement_height();
+
+
+  bool success = false;
```
Comment (tewinget) [inline src/blockchain_db/sqlite/db_sqlite.cpp:426]: same as above, `success` not actually being used, just looks like it is.

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 36]
```diff
@@ -0,0 +1,666 @@
...
+  try
+  {
+    SQLite::Transaction transaction{db, SQLite::TransactionBehavior::IMMEDIATE};
+
+
+    // Step 1: Remove the block producers txn fees
+    uint64_t service_node_reward = cryptonote::service_node_reward_formula(0, block.major_version);
+    if (block.reward - service_node_reward > 0)
```
Comment (jagerman) [inline src/blockchain_db/sqlite/db_sqlite.cpp:434]:
```suggestion
    if (block.reward - service_node_reward > 0) // If true then we have tx fees to roll back
```

File: src/blockchain_db/sqlite/db_sqlite.cpp (hunk 28 above)
Comment (jagerman) [inline src/blockchain_db/sqlite/db_sqlite.cpp:436]:
```suggestion
      if (block.service_node_winner_key // "service_node_winner_key" tracks the pulse winner; 0 if a mined block
        && crypto_core_ed25519_is_valid_point(reinterpret_cast<const unsigned char *>(block.service_node_winner_key.data)))
```

Comment (jagerman) [reply src/blockchain_db/sqlite/db_sqlite.cpp:436]: It does.

File: src/blockchain_db/sqlite/db_sqlite.cpp [hunk 37]
```diff
@@ -0,0 +1,666 @@
...
+      if (payable_service_node == service_nodes_state.service_nodes_infos.end())
+        continue;
+      std::vector<cryptonote::batch_sn_payment> node_rewards = calculate_rewards(block.major_version, service_node_reward/total_service_nodes_payable, *payable_service_node->second);
+      // Takes the node and adds its contributors to the batching database
+      success = subtract_sn_payments(node_rewards);
+      if (!success)
+        return false;
+    }
+    // Step 3: Remove Governance reward
+    if (m_nettype != cryptonote::FAKECHAIN)
```
Comment (jagerman) [inline src/blockchain_db/sqlite/db_sqlite.cpp:466]: Why this?  Does fakechain not have a governance reward?

File: src/blockchain_db/sqlite/db_sqlite.cpp (hunk 32 above)
Comment (darcys22) [reply src/blockchain_db/sqlite/db_sqlite.cpp:310]: have removed almost all of these. There were a couple edge cases in tests that cause sync issues so had to keep them in (Unit test where genesis block == hardfork, forking the chaingen struct which maintains its own batching database).

File: src/blockchain_db/sqlite/db_sqlite.cpp (hunk 37 above)
Comment (darcys22) [reply src/blockchain_db/sqlite/db_sqlite.cpp:466]: Yeah the tests dont ever pay a governance reward. They just want to add blocks with whatever they want and not worry about it.
